lib/eventsource/index.js

```diff
@@ -0,0 +1,448 @@
+      mode: 'cors',
+      credentials: corsAttributeState === 'anonymous'
+        ? 'same-origin'
+        : 'omit',
+      referrer: 'no-referrer',
+      referrerPolicy: 'no-referrer'
+    }
+
+    // 9. Set request's client to settings.
+
+    // 10. User agents may set (`Accept`, `text/event-stream`) in request's header list.
+    initRequest.headers = new HeadersList()
+    initRequest.headers.set('accept', 'text/event-stream', true)
+
+    // 11. Set request's cache mode to "no-store".
+    initRequest.cache = 'no-store'
+
+    // 12. Set request's initiator type to "other".
+    initRequest.initiator = 'other'
+
+    initRequest.urlList = [new URL(this.#url)]
+
+    // 13. Set ev's request to request.
+    this.#request = makeRequest(initRequest)
+
+    this.#request.headersList = initRequest.headers
+
+    this.#connect()
+  }
+
+  /**
+   * Returns the state of this EventSource object's connection. It can have the
+   * values described below.
+   * @returns {0|1|2}
+   * @readonly
+   */
+  get readyState () {
+    return this.#readyState
+  }
+
+  /**
+   * Returns the URL providing the event stream.
+   * @readonly
+   * @returns {string}
+   */
+  get url () {
+    return this.#url
+  }
+
+  /**
+   * Returns a boolean indicating whether the EventSource object was
+   * instantiated with CORS credentials set (true), or not (false, the default).
+   */
+  get withCredentials () {
+    return this.#withCredentials
+  }
+
+  async #connect () {
+    if (this.#readyState === CLOSED) return
+
+    this.#readyState = CONNECTING
+
+    const fetchParam = {
+      request: this.#request
+    }
+
+    // 14. Let processEventSourceEndOfBody given response res be the following step: if res is not a network error, then reestablish the connection.
+    const processEventSourceEndOfBody = (response) => {
+      if (isNetworkError(response)) {
+        this.dispatchEvent(new ErrorEvent('error', { error: response.error }))
+        this.close()
+      }
+
+      this.#connect()
+    }
+
+    // 15. Fetch request, with processResponseEndOfBody set to processEventSourceEndOfBody...
+    fetchParam.processResponseEndOfBody = processEventSourceEndOfBody
+
+    // and processResponse set to the following steps given response res:
+
+    fetchParam.processResponse = (response) => {
+      // 1. If res is an aborted network error, then fail the connection.
+
+      if (isNetworkError(response)) {
+        // 1. When a user agent is to fail the connection, the user agent
+        // must queue a task which, if the readyState attribute is set to a
+        // value other than CLOSED, sets the readyState attribute to CLOSED
+        // and fires an event named error at the EventSource object. Once the
+        // user agent has failed the connection, it does not attempt to
+        // reconnect.
+        if (response.aborted) {
+          this.close()
+          this.dispatchEvent(new ErrorEvent('error', { error: response.error }))
```

Comment:
I think I mentioned it somewhere but is this meant to be an ErrorEvent, or a `new Event('error')`? The spec text makes it seem like the latter.

Reply:
This was regarding OpenEvent. I removed that class. ErrorEvent was already existing in websocket. I think i forgot to remove the types for OpenEvent.

Reply:
From what I remember the spec uses the same language "an event named [open/error]". I checked Deno's code and it seemed to use `new Event("error")`.